### PR TITLE
Introduce AppKey for type safety

### PR DESCRIPTION
**Pull request by Dreamsorcerer**
Will create some PRs in other projects shortly to test it all out properly. But, this looks promising so far. It's less intrusive than the previous proposal and still works with `Request.config_dict`.

### CODE REVIEW

File: aiohttp/helpers.py [hunk 1]
```diff
@@ -784,11 +786,41 @@ def set_exception(fut: "asyncio.Future[_T]", exc: BaseException) -> None:
         fut.set_exception(exc)
 
 
+@functools.total_ordering
+class AppKey(Generic[_T]):
```
Comment (Dreamsorcerer) [inline aiohttp/helpers.py:795]: I initially put this in web_app.py, but that looks like a circular import to reference it here. Maybe I can still move it back though and do a `if TYPE_CHECKING` thing here.

Comment (asvetlov) [reply aiohttp/helpers.py:795]: I think this is fine

**Comment by asvetlov** [comment]
If it makes linter problems, we can drop inheritance from `Mapping` / `MutableMapping`.
Strictly speaking, the proposed structure is a heterogeneous container; `typing.Mapping[K, V]` is more likely a homogenous version.
Mypy should still be smart enough to infer types from `__getitem__` / `__setitem__`, isn't it?
The support of runtime `collections.abc` check is still desirable. It can be done either by inheritance or by direct `collections.abc.mapping.register()` call -- both works to me.

**Comment by Dreamsorcerer** [comment]
> If it makes linter problems, we can drop inheritance from `Mapping` / `MutableMapping`.

I'll just drop a `type: ignore` in, should work otherwise. I was just hoping to get a better approach from https://github.com/python/mypy/issues/11949

**Comment by Dreamsorcerer** [comment]
So, it doesn't work as nicely as I'd like.

`AppKey("foo", Iterable[int])` won't work as it's not a concrete class. So, instead we probably have to do it like the Stash implementation as `AppKey[Iterable[int]]("foo")`.

Another issue is that the typing errors from mypy are rather difficult to work with. For example, this code in aiohttp-jinja2:
```
app[aiohttp_jinja2.APP_CONTEXT_PROCESSORS_KEY] = (
    aiohttp_jinja2.request_processor,
    processor,
)
```
Produces:
```
Cannot infer type argument 1 of "__setitem__" of "Application"  [misc]
```

But, it works if you write it as:
```
f: Tuple[aiohttp_jinja2._ContextProcessor, ...] = (
    aiohttp_jinja2.request_processor,
    processor,
)
app[aiohttp_jinja2.APP_CONTEXT_PROCESSORS_KEY] = f
```


Even with simple examples like:
```
k = web.AppKey[int]("k")
app[k] = "foo"
```
Mypy just produces: `Cannot infer type argument 1 of "__setitem__" of "Application"  [misc]`

Which is rather awkward for the user to figure out the problem. The type of both items is defined as a `TypeVar` (`def __setitem__(self, key: AppKey[_T], value: _T)`), so mypy only tells you if they are not the same type, but it doesn't tell you that the value type is the problem or what the expected type is.

**Comment by Dreamsorcerer** [comment]
Anyway, changes for aiohttp-jinja2: https://github.com/aio-libs/aiohttp-jinja2/pull/534/files

**Comment by Dreamsorcerer** [comment]
> `AppKey("foo", Iterable[int])` won't work as it's not a concrete class.

Weirdly, mypy stop complaining about this once the parameter is defined with `Optional`. I'd rather like to keep this syntax, so I've added support for both.

**Review by asvetlov — APPROVED** [review]
LGTM.
@Dreamsorcerer please feel free to merge when you decide that the PR is ready.

**Comment by Dreamsorcerer** [comment]
Right, I'm going to merge this and hope that mypy improves its error message in future. I opened an issue for it: https://github.com/python/mypy/issues/12012

**Comment by eoghanmurray** [comment]
What if I don't care about types (because I'm using Python, not Typescript)? 
I want to do 
```
app['redis'] = RedisCluster()
app['port'] = 30000
```
Without having to figure out why my app is warning about AppKeys etc.?
Specifically IMO the warning `UserWarning: It is recommended to use web.AppKey instances for keys.` would be an appropriate warning given by mypy or some tool as it is running type checking over an app, it is inappropriate for an application startup warning.

**Comment by Dreamsorcerer** [comment]
Firstly, it's a warning, just filter it out if you decide it's not relevant for you. Secondly, it's not possible to create a warning in mypy for something like this (especially retroactively, while maintaining backwards-compatibility).

I'll add a note to that section with examples to filter out the warning, but it's just a standard warning. We added a custom class, so you can filter out that specific warning very easily.

**Comment by Dreamsorcerer** [comment]
Also, even if you don't use mypy, a lot of users still want annotations etc. in their IDE via pyright or similar.

**Comment by Dreamsorcerer** [comment]
https://aiohttp--7839.org.readthedocs.build/en/7839/web_advanced.html#application-s-config

**Comment by eoghanmurray** [comment]
Thanks for your response!

> Just filter it out

To do that I have to figure out what the warning means, and whether it's something that I need to fix. In this case the fix (using a WebKey class when both setting and retrieving, and figuring out types for values which would then need to be maintained) seems quite un-pythonic and would make the code more mysterious.

I did already try to filter it out before posting here (I didn't previously know if filtering warnings was possible in Python) but the following code taken from SO didn't work

```
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    app['redis'] = RedisCluster()
    app['port'] = 30000
```

As I now have learned that you need to set up the 'ignore' before `app = web.Application()` is initialized.

I just think that this is something that should be highlighted as best practices in the documentation and doesn't merit a warning.
With thanks for your time.

**Comment by Dreamsorcerer** [comment]
```
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    app["foo"] = 3
```

Works fine for  me.

Anyway, if you don't care at all, filter it globally, examples in the above link. I'd use catch_warnings() if you're only trying to maintain backwards-compatibility with something (which is likely what homeassistant will be doing).

**Comment by eoghanmurray** [comment]
My version is 3.9.0b0 which doesn't appear to have #7677 in it "Only display warning when using -Wdefault, -We, -X dev or similar."

So basically my 'observations' are redundant I think (I didn't have warnings explicitly turned on).

**Comment by Dreamsorcerer** [comment]
Ah, yes, I didn't realise your were on b0. I did mention that the warning shouldn't appear unless you're explicitly enabling them with `-Wall` or similar. That was a bug, as I thought it would be ignored by default.

rc0 is building now.

**Comment by eoghanmurray** [comment]
I apologize for wasting your time, `3.9.0b1` works perfect in this respect.

**Pull request merged**